File: internal/match/request.go
// SPDX-License-Identifier: Apache-2.0

package match

import (
	"net/http"
	"net/url"
	"strings"
	"sync"

	"github.com/b3vet/mockulus/internal/matchers"
)

// ParsedRequest is the per-request working set: the request as the matcher
// needs to see it, with every derived form computed at most once and only when
// something asks for it.
//
// The laziness is the point. A stub matched on method and URL alone never
// causes its request's body to be read as a string, parsed as JSON, or split
// into form fields — features are pay-per-use, and a deployment that uses none
// of them pays for none of them (SPEC §6.4, P2).
//
// Instances come from a pool and are reused. Reset drops every reference so no
// request-scoped memory outlives the request that filled it.
type ParsedRequest struct {
	// Method is the request method, upper case.
	Method string
	// Path is the request path with no query string.
	Path string
	// FullURL is path and query exactly as received, which is what a byte-exact
	// `url` criterion compares against.
	FullURL string

	header http.Header
	body   []byte

	query       url.Values
	queryParsed bool

	cookies       map[string][]string
	cookiesParsed bool

	form       url.Values
	formParsed bool

	// bodySubject persists for the whole request so a body examined by several
	// matchers is converted and parsed once.
	bodySubject matchers.Body

	// keyScratch backs the subject handed to each key criterion. One suffices:
	// criteria are evaluated strictly one at a time.
	keyScratch matchers.KeyValues

	// keyBuf assembles the snapshot index keys, whose shape is
	// "METHOD\x00url" (SPEC §6.1). See indexLookup for why they are not simply
	// concatenated.
	keyBuf []byte

	// pathVars holds the bindings produced by a urlPathTemplate match, which
	// the pathParameters criteria of that same stub then consume.
	pathVars map[string]string

	// scenarioStates memoizes scenario state reads so a request touching
	// several stubs in one scenario reads its state once (SPEC §9.2).
	scenarioStates map[string]string
	// scenarioErr holds a state read that could not be answered. It is carried
	// out of matching rather than returned from it because the gate is a
	// predicate: without somewhere to put the failure, an unreachable store
	// reads as "this stub does not match" and the request is answered from the
	// wrong side of the state machine (SPEC §9.2).
	scenarioErr error
}

var requestPool = sync.Pool{
	New: func() any {
		return &ParsedRequest{
			pathVars:       make(map[string]string, 4),
			scenarioStates: make(map[string]string, 2),
			keyBuf:         make([]byte, 0, initialKeyBuf),
		}
	},
}

// initialKeyBuf comfortably holds a method and a path of ordinary length, so a
// pooled request assembles its index keys without growing the buffer.
const initialKeyBuf = 128

// maxPooledKeyBuf bounds what a pooled request keeps. The buffer grows to the
// longest URL that instance has seen, and net/http will accept a request line
// close to a megabyte; without this, one such request permanently inflates a
// pool entry that every later request reuses.
const maxPooledKeyBuf = 4 << 10

// AcquireRequest takes a ParsedRequest from the pool and binds it to an
// incoming request and its already-read body.
func AcquireRequest(r *http.Request, body []byte) *ParsedRequest {
	pr, _ := requestPool.Get().(*ParsedRequest)
	pr.bind(r, body)
	return pr
}

// ReleaseRequest returns a ParsedRequest to the pool after clearing it.
func ReleaseRequest(pr *ParsedRequest) {
	pr.Reset()
	requestPool.Put(pr)
}

func (r *ParsedRequest) bind(req *http.Request, body []byte) {
	r.Method = req.Method
	r.FullURL = requestTarget(req)
	r.Path = r.FullURL
	if i := strings.IndexByte(r.FullURL, '?'); i >= 0 {
		r.Path = r.FullURL[:i]
	}
	r.header = req.Header
	r.body = body
	r.bodySubject.Set(body)
}

// requestTarget returns the request line's target: path and query exactly as
// the client wrote them, which is what a byte-exact `url` criterion compares
// against.
//
// net/http keeps that string verbatim in RequestURI, and re-deriving it from
// the parsed URL instead is not free — URL.RequestURI re-escapes the path,
// scanning it byte by byte with a per-byte function call, which the serve
// profile showed costing as much as the whole match. For an origin-form target
// the two agree exactly, because url.setPath keeps the raw form in RawPath
// whenever the default encoding would differ from it;
// TestRequestTargetIsTheTargetAsReceived holds them to that over the encodings
// where it would be easiest to be wrong.
//
// The leading-slash guard is what limits the shortcut to the origin form. A
// proxy client's absolute target is the case where the two genuinely disagree —
// RequestURI carries the scheme and host, and a stub's URL criterion is written
// against neither — so that one goes the long way round.
func requestTarget(req *http.Request) string {
	if uri := req.RequestURI; uri != "" && uri[0] == '/' {
		return uri
	}
	return req.URL.RequestURI()
}

// Reset clears every reference the request held. Pooled memory that keeps a
// pointer to a previous request's body is a leak that only shows up under load,
// so this is deliberately exhaustive.
func (r *ParsedRequest) Reset() {
	r.Method, r.Path, r.FullURL = "", "", ""
	r.header = nil
	r.body = nil

	r.query, r.queryParsed = nil, false
	r.cookies, r.cookiesParsed = nil, false
	r.form, r.formParsed = nil, false

	r.bodySubject.Reset()
	r.keyScratch.Set(false, nil)

	// Truncated rather than dropped: it holds bytes copied out of the request
	// rather than a reference into it, and keeping the capacity is the whole
	// reason index lookups allocate nothing. An outsized one is dropped.
	if cap(r.keyBuf) > maxPooledKeyBuf {
		r.keyBuf = nil
	} else {
		r.keyBuf = r.keyBuf[:0]
	}

	clear(r.pathVars)
	clear(r.scenarioStates)
	r.scenarioErr = nil
}

// Body returns the raw request body.
func (r *ParsedRequest) Body() []byte { return r.body }

// indexLookup probes one of the snapshot's URL indexes for this request.
//
// The key is assembled into the request's own buffer rather than written as
// method+methodSep+url, because Go keeps a concatenated string off the heap
// only while it fits a 32-byte stack buffer. A method plus any realistic path
// exceeds that, and every request probes four keys — so the obvious spelling
// costs four heap allocations on the exact-URL hit that most traffic is, while
// a benchmark using a short path reports none of them (SPEC §16.3 rule 1). The
// map lookup takes a string view of the buffer, which the compiler does not
// copy.
func (r *ParsedRequest) indexLookup(index map[string][]int32, method, url string) []int32 {
	r.keyBuf = append(r.keyBuf[:0], method...)
	r.keyBuf = append(r.keyBuf, methodSep...)
	r.keyBuf = append(r.keyBuf, url...)
	return index[string(r.keyBuf)]
}

// HeaderValues returns every value sent for a header, looked up
// case-insensitively.
func (r *ParsedRequest) HeaderValues(name string) []string { return r.header.Values(name) }

// BodySubject returns the subject for body criteria.
func (r *ParsedRequest) BodySubject() matchers.Subject { return &r.bodySubject }

// HeaderSubject returns the subject for a header, matched case-insensitively
// as WireMock does. Go canonicalises header names on both store and lookup, so
// the case-insensitivity comes for free and correctly.
func (r *ParsedRequest) HeaderSubject(name string) matchers.Subject {
	values := r.header.Values(name)
	r.keyScratch.Set(values != nil, values)
	return &r.keyScratch
}

// QuerySubject returns the subject for a query parameter. Parameter names are
// case-sensitive, unlike header names.
func (r *ParsedRequest) QuerySubject(name string) matchers.Subject {
	if !r.queryParsed {
		r.query = parseQuery(r.FullURL)
		r.queryParsed = true
	}
	values, present := r.query[name]
	r.keyScratch.Set(present, values)
	return &r.keyScratch
}

// CookieSubject returns the subject for a cookie.
func (r *ParsedRequest) CookieSubject(name string) matchers.Subject {
	if !r.cookiesParsed {
		r.cookies = parseCookies(r.header.Values("Cookie"))
		r.cookiesParsed = true
	}
	values, present := r.cookies[name]
	// An absent cookie fails a negative matcher too, unlike an absent header.
	r.keyScratch.SetStrictAbsence(present, values)
	return &r.keyScratch
}

// FormSubject returns the subject for a form field, parsing the body as
// urlencoded form data on first use. A request that is not form-encoded has no
// form fields rather than an error: the criterion simply does not match.
//
// The media type is read here rather than in bind, because this is the only
// thing that wants it: deriving it up front would charge every request for a
// header lookup and a case fold on behalf of the one stub in a thousand that
// matches on form parameters (P2).
func (r *ParsedRequest) FormSubject(name string) matchers.Subject {
	if !r.formParsed {
		r.formParsed = true
		if mediaType(r.header.Get("Content-Type")) == "application/x-www-form-urlencoded" {
			if parsed, err := url.ParseQuery(string(r.body)); err == nil {
				r.form = parsed
			}
		}
	}
	values, present := r.form[name]
	r.keyScratch.Set(present, values)
	return &r.keyScratch
}

// PathVarSubject returns the subject for a path-template variable bound by the
// URL match of the stub currently being evaluated.
func (r *ParsedRequest) PathVarSubject(name string) matchers.Subject {
	value, present := r.pathVars[name]
	if !present {
		r.keyScratch.Set(false, nil)
		return &r.keyScratch
	}
	r.keyScratch.Set(true, []string{value})
	return &r.keyScratch
}

// PathVars exposes the bindings of the stub that matched, which a response
// template reads as request.path.<name>.
func (r *ParsedRequest) PathVars() map[string]string { return r.pathVars }

// BindPathVar records a path-template binding for the stub being evaluated.
func (r *ParsedRequest) BindPathVar(name, value string) { r.pathVars[name] = value }

// ClearPathVars drops the bindings of the previously evaluated stub, so one
// stub's variables can never leak into another's criteria.
func (r *ParsedRequest) ClearPathVars() { clear(r.pathVars) }

// ScenarioState returns a memoized scenario state read.
func (r *ParsedRequest) ScenarioState(name string) (string, bool) {
	state, ok := r.scenarioStates[name]
	return state, ok
}

// MemoizeScenarioState records a state read so the same request does not read
// it again (SPEC §9.2).
func (r *ParsedRequest) MemoizeScenarioState(name, state string) {
	r.scenarioStates[name] = state
}

// FailScenarioRead records a state read the store could not answer.
//
// The first failure is kept: it is the one that made the request unanswerable,
// and a later candidate's error says nothing new about why.
func (r *ParsedRequest) FailScenarioRead(err error) {
	if r.scenarioErr == nil {
		r.scenarioErr = err
	}
}

// ScenarioError returns the state read that failed, or nil.
func (r *ParsedRequest) ScenarioError() error { return r.scenarioErr }

// parseQuery extracts the query string from a request URI and parses it.
// Values that fail to unescape are kept in raw form rather than dropped: a stub
// matching on an odd-looking parameter should still see it.
func parseQuery(requestURI string) url.Values {
	i := strings.IndexByte(requestURI, '?')
	if i < 0 {
		return url.Values{}
	}
	values, err := url.ParseQuery(requestURI[i+1:])
	if err != nil && values == nil {
		return url.Values{}
	}
	return values
}

// parseCookies splits Cookie header values into name/value pairs. A name may
// legitimately appear more than once, so values accumulate.
func parseCookies(headers []string) map[string][]string {
	if len(headers) == 0 {
		return nil
	}
	out := make(map[string][]string, 4)
	for _, header := range headers {
		for _, pair := range strings.Split(header, ";") {
			pair = strings.TrimSpace(pair)
			if pair == "" {
				continue
			}
			name, value, found := strings.Cut(pair, "=")
			if !found {
				continue
			}
			name = strings.TrimSpace(name)
			value = strings.Trim(strings.TrimSpace(value), `"`)
			out[name] = append(out[name], value)
		}
	}
	return out
}

// mediaType strips parameters and case from a Content-Type header, so
// "application/json; charset=utf-8" compares as "application/json".
func mediaType(header string) string {
	if header == "" {
		return ""
	}
	if i := strings.IndexByte(header, ';'); i >= 0 {
		header = header[:i]
	}
	return strings.ToLower(strings.TrimSpace(header))
}

File: sdk/typescript/src/builders/request.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * The request half of a stub, and the mapping the two halves make together —
 * under the names WireMock's Java DSL gives them.
 *
 * A request matches when **every** stated criterion matches, and an absent
 * criterion is a wildcard, so a stub built from `get(anyUrl())` alone covers
 * every GET. That is the server's rule and WireMock's, and it is why nothing
 * here has a default: a criterion that was not written is not a criterion.
 *
 * Three refusals the server would answer with a 422 are unwritable here
 * instead, and each one is a shape WireMock accepts and then behaves oddly
 * around:
 *
 * - **Two URL criteria.** The verb takes exactly one, so there is no second to
 *   give. WireMock resolves a pair by a fixed field precedence and its echo
 *   silently omits the criteria it discarded, so a stub matching on a field its
 *   author did not intend reads back as though the others were never written
 *   (deviation #47).
 * - **A path-parameter criterion with no template to bind against, or naming a
 *   variable the template does not bind.** {@link urlPathTemplate} carries its
 *   variable names in its type, so `withPathParam` accepts those names and
 *   nothing else, and does not exist at all on a stub whose URL criterion is
 *   not a template. WireMock accepts the first case and drops the whole block,
 *   so an unsatisfiable criterion registers and the stub then matches *every*
 *   request — the widest possible failure, arrived at silently (deviation #54).
 * - **A scenario state with no scenario.** `whenScenarioStateIs` and
 *   `willSetStateTo` appear only after `inScenario`, which is the shape
 *   WireMock's own Java builder has.
 */

import type { RequestPattern, StubMapping } from '../types.js';
import type { BodyPattern, Matcher } from './matchers.js';
import type { ResponseBuilder } from './response.js';

/**
 * The variable names a path template binds.
 *
 * A variable is a whole segment, so `/orders/{id}` binds `id` and `/a{b}c` is
 * refused by {@link urlPathTemplate} before this type is ever consulted — which
 * is why the extraction below can be the simple one.
 */
export type PathVariables<Template extends string> =
  Template extends `${string}{${infer Name}}${infer Rest}` ? Name | PathVariables<Rest> : never;

/** Whether a mapping has named a scenario, and so whether it may name a state. */
type ScenarioState = 'none' | 'named';

/** The five ways a request pattern may state a URL criterion. */
type UrlField = keyof Pick<
  RequestPattern,
  'url' | 'urlPattern' | 'urlPath' | 'urlPathPattern' | 'urlPathTemplate'
>;

/**
 * One URL criterion, and the path variables it binds.
 *
 * It is a class rather than a document fragment so that a caller cannot write a
 * second URL field into it: the only way to make one is through the six
 * functions below, and each of them writes exactly one field. That is the
 * mutual exclusion the contract declines to express — "a mutually exclusive
 * union of five variants would make the shared type unusable for the SDK" — put
 * where a builder can carry it.
 */
export class UrlCriterion<Variables extends string = never> {
  /**
   * Present so `UrlCriterion<'id'>` and `UrlCriterion<never>` are different
   * types. `declare` because the names live in {@link PathVariables} and are
   * never needed at runtime.
   */
  declare private readonly variables: Variables;

  /** @param field The one field this criterion writes; `undefined` for {@link anyUrl}. */
  constructor(
    readonly field: UrlField | undefined,
    readonly value: string,
  ) {}

  /** The fragment this criterion contributes to a request pattern. */
  toFragment(): RequestPattern {
    return this.field === undefined ? {} : { [this.field]: this.value };
  }
}

/**
 * **Byte-exact** match on path and query as received, so query parameter order
 * matters: `?a=1&b=2` and `?b=2&a=1` are different criteria and only one of them
 * matches a given request.
 */
export function urlEqualTo(url: string): UrlCriterion {
  return new UrlCriterion('url', absolutePath(url, 'urlEqualTo'));
}

/** A regular expression matched in full against path and query together. */
export function urlMatching(pattern: string): UrlCriterion {
  return new UrlCriterion('urlPattern', pattern);
}

/** Exact match on the path alone, ignoring the query. */
export function urlPathEqualTo(path: string): UrlCriterion {
  return new UrlCriterion('urlPath', absolutePath(path, 'urlPathEqualTo'));
}

/** A regular expression matched in full against the path alone. */
export function urlPathMatching(pattern: string): UrlCriterion {
  return new UrlCriterion('urlPathPattern', pattern);
}

/**
 * A WireMock 3 path template such as `/orders/{id}`, matched against the path
 * and binding each variable for {@link MappingBuilder.withPathParam}.
 *
 * The template is checked here, by the same rules the server applies, because
 * the type below is derived from it: a template the server would refuse would
 * otherwise produce a set of variable names that is not the set the stub ends up
 * binding, and the compile-time guarantee would be over a template that never
 * registers.
 */
export function urlPathTemplate<Template extends string>(
  template: Template,
): UrlCriterion<PathVariables<Template>> {
  return new UrlCriterion<PathVariables<Template>>('urlPathTemplate', checkedTemplate(template));
}

/**
 * No URL criterion at all, which is a wildcard.
 *
 * WireMock spells it `anyUrl()` and means the same thing: the field is absent
 * from the document rather than present and empty.
 */
export function anyUrl(): UrlCriterion {
  return new UrlCriterion(undefined, '');
}

/**
 * A stub under construction.
 *
 * The builder is immutable — every call answers a new one — so a partly-built
 * mapping can be shared between cases and specialised two ways without the
 * second inheriting the first's criteria. WireMock's Java builder mutates in
 * place, and this is the one place these builders deliberately differ from it.
 *
 * `Variables` is the set of names the URL template binds, and `Scenario`
 * records whether a scenario has been named. Both exist only in the type
 * system.
 */
export class MappingBuilder<
  Variables extends string = never,
  Scenario extends ScenarioState = 'none',
> {
  /** Present so two instantiations are different types — see {@link UrlCriterion}. */
  declare private readonly variables: Variables;
  /** Present for the same reason, so the scenario methods can require a scenario. */
  declare private readonly scenario: Scenario;

  private constructor(private readonly document: StubMapping) {}

  /** Starts a mapping from a method and a URL criterion. Not the entry point; the verbs are. */
  static begin<Variables extends string>(
    method: string,
    url: UrlCriterion<Variables>,
  ): MappingBuilder<Variables> {
    return new MappingBuilder<Variables>({ request: { method, ...url.toFragment() } });
  }

  /** The mapping as it stands. {@link stubFor} is the spelling a caller uses. */
  toDocument(): StubMapping {
    return { ...this.document };
  }

  /**
   * A criterion over one request header.
   *
   * Header names are case-insensitive in both directions, so a criterion on
   * `Accept` matches a request that sent `accept`. Values are case-sensitive
   * unless the matcher folds case.
   *
   * A repeated header matches when **any** of its values satisfies the matcher.
   * WireMock instead picks the value at minimum edit distance and matches that
   * one, so mockulus matches strictly more here and no suite that passes on
   * WireMock can fail on this (deviation #29).
   */
  withHeader(name: string, matcher: Matcher): MappingBuilder<Variables, Scenario> {
    return this.withKeyCriterion('headers', name, matcher);
  }

  /**
   * A criterion over one query parameter.
   *
   * A repeated parameter matches when any of its values satisfies the matcher.
   * `?x=` and a bare `?x` are both present-with-empty-string, never absent, so
   * `absent()` is the only way to say "must not be present" and `equalTo('')` is
   * how the two empty spellings are matched.
   */
  withQueryParam(name: string, matcher: Matcher): MappingBuilder<Variables, Scenario> {
    return this.withKeyCriterion('queryParameters', name, matcher);
  }

  /** A criterion over one cookie. */
  withCookie(name: string, matcher: Matcher): MappingBuilder<Variables, Scenario> {
    return this.withKeyCriterion('cookies', name, matcher);
  }

  /**
   * A criterion over one field of an `application/x-www-form-urlencoded` body.
   *
   * The body is parsed lazily, so a request that fails on URL or method never
   * pays for the parse.
   */
  withFormParam(name: string, matcher: Matcher): MappingBuilder<Variables, Scenario> {
    return this.withKeyCriterion('formParameters', name, matcher);
  }

  /**
   * A criterion over one variable of the URL template.
   *
   * The name is one of the template's own variables — the type says so — which
   * is what makes the two refusals around this field unwritable rather than
   * merely documented. On a stub whose URL criterion is not a template,
   * `Variables` is `never` and there is no name that can be passed at all.
   */
  withPathParam(name: Variables, matcher: Matcher): MappingBuilder<Variables, Scenario> {
    return this.withKeyCriterion('pathParameters', name, matcher);
  }

  /**
   * A criterion over the request body.
   *
   * Several of these are an AND: **all** of them must match. They are evaluated
   * cheapest first — equality before regex before anything that has to parse the
   * body — which is an ordering the server chooses and no stub needs to think
   * about, so the order they are written in is a matter of reading rather than
   * of cost.
   *
   * This is the one position that takes {@link BodyPattern} rather than
   * {@link Matcher}, because it is the one position where `binaryEqualTo` has
   * raw bytes to compare against.
   */
  withRequestBody(matcher: BodyPattern): MappingBuilder<Variables, Scenario> {
    return this.withRequest({
      bodyPatterns: [...(this.document.request?.bodyPatterns ?? []), matcher],
    });
  }

  /**
   * Sugar over an `Authorization` header criterion.
   *
   * Both halves are required because a half-written credential is a criterion
   * its author cannot have meant, and the server refuses one.
   */
  withBasicAuth(username: string, password: string): MappingBuilder<Variables, Scenario> {
    return this.withRequest({ basicAuthCredentials: { username, password } });
  }

  /** A human-readable label, stored, echoed and shown in near-miss output. */
  withName(name: string): MappingBuilder<Variables, Scenario> {
    return this.with({ name });
  }

  /**
   * The stub's identity, which the server otherwise generates.
   *
   * It must be the canonical 36-character spelling, which is checked here
   * because the alternative is a 422 on a value that was probably produced by
   * the wrong formatter. The dashless, `urn:uuid:` and brace-wrapped spellings
   * are refused by WireMock too; the 24-character base64 encoding of the raw
   * bytes is one WireMock accepts and silently rewrites, and refusing it is what
   * stops a client being handed back an id it did not choose (deviation #24).
   *
   * `id` is written rather than `uuid` — they are aliases and both are echoed —
   * because sending a single spelling reproduces WireMock exactly, and sending
   * both is only safe while they agree.
   */
  withId(id: string): MappingBuilder<Variables, Scenario> {
    if (!/^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/.test(id)) {
      throw new TypeError(`withId needs a canonical 36-character UUID, got ${JSON.stringify(id)}`);
    }
    return this.with({ id });
  }

  /**
   * Selection order among matching stubs, lower first. Absent means 5.
   *
   * An arbitrary signed integer compared numerically — no clamping and no range
   * validation, which is what the pinned WireMock does, so the common "1 is the
   * highest" phrasing describes a convention rather than a constraint. The range
   * checked here is the one the field is declared over, and nothing narrower.
   *
   * Equal priorities break on insertion sequence, newest first, and that
   * sequence is a cluster-global counter, so "most recently added wins" holds
   * across replicas (SPEC §5.3, §7.3).
   */
  withPriority(priority: number): MappingBuilder<Variables, Scenario> {
    if (!Number.isInteger(priority) || priority < -2147483648 || priority > 2147483647) {
      throw new RangeError(`withPriority needs a 32-bit integer, got ${String(priority)}`);
    }
    return this.with({ priority });
  }

  /**
   * Arbitrary JSON, stored and echoed untouched, and searchable through
   * `find-by-metadata` and `remove-by-metadata`.
   *
   * Tagging every stub a run creates with a run id, and cleaning up through
   * `remove-by-metadata`, is the discipline SPEC §1 asks of suites sharing a
   * deployment — the alternative is a global reset that takes another team's
   * stubs with it.
   */
  withMetadata(metadata: Record<string, unknown>): MappingBuilder<Variables, Scenario> {
    return this.with({ metadata: { ...this.document.metadata, ...metadata } });
  }

  /**
   * Asks for a durable document rather than one with a TTL.
   *
   * Absent or `false` stores a document with `ephemeral_stub_ttl` — 24 hours by
   * default — which `POST /__admin/mappings/reset` also sweeps. WireMock keeps
   * non-persistent stubs until the process restarts; a TTL is the documented
   * equivalent for a long-running cluster that never restarts (deviation #3).
   */
  persistent(persistent = true): MappingBuilder<Variables, Scenario> {
    return this.with({ persistent });
  }

  /**
   * The scenario this stub belongs to.
   *
   * A scenario exists because stubs name it, so this call is what brings one
   * into being, and deleting the last stub that names it is what ends it. It is
   * also what unlocks the two state methods below, which the server refuses
   * without it.
   */
  inScenario(scenarioName: string): MappingBuilder<Variables, 'named'> {
    return new MappingBuilder<Variables, 'named'>({ ...this.document, scenarioName });
  }

  /**
   * The state the stub matches in.
   *
   * In any other state it is treated as non-matching and iteration continues, so
   * a scenario-gated stub does not shadow the stubs behind it.
   */
  whenScenarioStateIs(
    this: MappingBuilder<Variables, 'named'>,
    requiredScenarioState: string,
  ): MappingBuilder<Variables, 'named'> {
    return this.withScenario({ requiredScenarioState });
  }

  /**
   * The state the scenario moves to after this stub serves.
   *
   * The transition is a compare-and-swap, so two concurrent requests cannot both
   * drive the same step (SPEC §9.3).
   */
  willSetStateTo(
    this: MappingBuilder<Variables, 'named'>,
    newScenarioState: string,
  ): MappingBuilder<Variables, 'named'> {
    return this.withScenario({ newScenarioState });
  }

  /**
   * What a matching request is served.
   *
   * A stub with no response at all serves 200 with an empty body, so this is for
   * saying anything more than that.
   */
  willReturn(response: ResponseBuilder<'unset' | 'set'>): MappingBuilder<Variables, Scenario> {
    return this.with({ response: response.toDocument() });
  }

  /** A new builder over this document plus some top-level fields. */
  private with(fields: StubMapping): MappingBuilder<Variables, Scenario> {
    return new MappingBuilder<Variables, Scenario>({ ...this.document, ...fields });
  }

  /**
   * {@link with} for the scenario methods, whose `this` constraint has narrowed
   * the receiver to a scenario-carrying builder and thereby lost the ability to
   * return `Scenario` unchanged. It is `'named'` in both, by construction.
   */
  private withScenario(
    this: MappingBuilder<Variables, 'named'>,
    fields: StubMapping,
  ): MappingBuilder<Variables, 'named'> {
    return new MappingBuilder<Variables, 'named'>({ ...this.document, ...fields });
  }

  /** A new builder over this document plus some request-pattern fields. */
  private withRequest(fields: RequestPattern): MappingBuilder<Variables, Scenario> {
    return this.with({ request: { ...this.document.request, ...fields } });
  }

  /** A new builder with one more criterion in one of the five keyed blocks. */
  private withKeyCriterion(
    block: 'headers' | 'queryParameters' | 'cookies' | 'formParameters' | 'pathParameters',
    name: string,
    matcher: Matcher,
  ): MappingBuilder<Variables, Scenario> {
    return this.withRequest({ [block]: { ...this.document.request?.[block], [name]: matcher } });
  }
}

/**
 * A mapping builder whatever its URL template and scenario say.
 *
 * {@link stubFor} takes one of these because it does not care: by the time a
 * mapping is being turned into a document, the constraints the two parameters
 * carry have already been checked at every call that could have broken them.
 */
export type AnyMappingBuilder = MappingBuilder<string, 'none' | 'named'>;

/** A stub for GET requests. */
export function get<Variables extends string>(
  url: UrlCriterion<Variables> | string,
): MappingBuilder<Variables> {
  return MappingBuilder.begin('GET', urlOf(url));
}

/** A stub for POST requests. */
export function post<Variables extends string>(
  url: UrlCriterion<Variables> | string,
): MappingBuilder<Variables> {
  return MappingBuilder.begin('POST', urlOf(url));
}

/** A stub for PUT requests. */
export function put<Variables extends string>(
  url: UrlCriterion<Variables> | string,
): MappingBuilder<Variables> {
  return MappingBuilder.begin('PUT', urlOf(url));
}

/** A stub for PATCH requests. */
export function patch<Variables extends string>(
  url: UrlCriterion<Variables> | string,
): MappingBuilder<Variables> {
  return MappingBuilder.begin('PATCH', urlOf(url));
}

/**
 * A stub for DELETE requests.
 *
 * WireMock's Java DSL calls this `delete`, which JavaScript cannot: `delete` is
 * an operator and not a name a binding can take. `del` is the abbreviation the
 * ecosystem already uses for exactly this collision.
 */
export function del<Variables extends string>(
  url: UrlCriterion<Variables> | string,
): MappingBuilder<Variables> {
  return MappingBuilder.begin('DELETE', urlOf(url));
}

/** A stub for HEAD requests. */
export function head<Variables extends string>(
  url: UrlCriterion<Variables> | string,
): MappingBuilder<Variables> {
  return MappingBuilder.begin('HEAD', urlOf(url));
}

/** A stub for OPTIONS requests. */
export function options<Variables extends string>(
  url: UrlCriterion<Variables> | string,
): MappingBuilder<Variables> {
  return MappingBuilder.begin('OPTIONS', urlOf(url));
}

/** A stub for TRACE requests. */
export function trace<Variables extends string>(
  url: UrlCriterion<Variables> | string,
): MappingBuilder<Variables> {
  return MappingBuilder.begin('TRACE', urlOf(url));
}

/** A stub for requests of any method. */
export function any<Variables extends string>(
  url: UrlCriterion<Variables> | string,
): MappingBuilder<Variables> {
  return MappingBuilder.begin('ANY', urlOf(url));
}

/**
 * A stub for a method the functions above do not name.
 *
 * Any method name is accepted, including one WireMock's own enumeration does not
 * define, because a mock server exists to stand in for services that have them.
 * The name is compared after trimming and upper-casing, so `any` and `ANY` are
 * one criterion.
 */
export function request<Variables extends string>(
  method: string,
  url: UrlCriterion<Variables> | string,
): MappingBuilder<Variables> {
  return MappingBuilder.begin(method, urlOf(url));
}

/**
 * Turns a mapping builder into the document to register.
 *
 * WireMock's `stubFor` registers with a server; this answers the mapping, and
 * `client.mappings.create` is what registers it. The split is deliberate: this
 * package has a client with a base URL, a token and a timeout, and a builder
 * that reached past it to a global would be a second, invisible way to configure
 * where a stub goes.
 */
export function stubFor(builder: AnyMappingBuilder): StubMapping {
  return builder.toDocument();
}

/** Reads the shorthand a verb accepts: a bare string is an exact-URL criterion, as in WireMock. */
function urlOf<Variables extends string>(
  url: UrlCriterion<Variables> | string,
): UrlCriterion<Variables> {
  return typeof url === 'string' ? (urlEqualTo(url) as UrlCriterion<Variables>) : url;
}

/**
 * Checks the two URL criteria the server requires to be absolute.
 *
 * `url` and `urlPath` are compared against the path as received, which always
 * starts with a slash, so one that does not could never match anything. The
 * pattern forms are not checked: a regular expression legitimately starts with
 * an anchor, a group or a character class.
 */
function absolutePath(path: string, caller: string): string {
  if (!path.startsWith('/')) {
    throw new TypeError(`${caller} needs a path starting with /, got ${JSON.stringify(path)}`);
  }
  return path;
}

/**
 * Checks a path template by the rules the server parses one with.
 *
 * A variable is a whole segment, no variable is bound twice, and the template
 * starts with a slash. Checking here rather than leaving it to registration is
 * what makes {@link PathVariables} trustworthy: the type reads the braces out of
 * the literal, and if the server would have read them differently the set of
 * names the compiler enforces would not be the set the stub binds.
 */
function checkedTemplate(template: string): string {
  if (!template.startsWith('/')) {
    throw new TypeError(
      `urlPathTemplate needs a template starting with /, got ${JSON.stringify(template)}`,
    );
  }
  const bound = new Set<string>();
  for (const segment of template.slice(1).split('/')) {
    if (segment.startsWith('{') && segment.endsWith('}')) {
      const name = segment.slice(1, -1);
      if (name === '' || /[{}/]/.test(name)) {
        throw new TypeError(`urlPathTemplate has a malformed variable ${JSON.stringify(segment)}`);
      }
      if (bound.has(name)) {
        throw new TypeError(`urlPathTemplate binds ${JSON.stringify(name)} more than once`);
      }
      bound.add(name);
      continue;
    }
    if (/[{}]/.test(segment)) {
      // A brace outside a whole-segment variable is a typo rather than a
      // template WireMock would accept, and saying so is better than matching it
      // literally and leaving the author puzzled about why nothing binds.
      throw new TypeError(
        `urlPathTemplate needs a variable to be a whole segment, as in /orders/{id}, ` +
          `got ${JSON.stringify(template)}`,
      );
    }
  }
  return template;
}
